services/tables/src/test/resources/cluster-test-properties.yaml

```
@@ -2,6 +2,21 @@ cluster:
   name: "TestCluster"
   storage:
     root-path: "/tmp/unittest"
+  storages:
+    default-type: "hdfs"
+    types:
+      hdfs:
+        scheme: "hdfs"
```

```suggestion
        type: "hdfs"
```
scheme has a very specific meaning. cloud blob stores like s3 can support an endpoint like 
`s3:///` where scheme is s3, and `https://`

I think we should rename scheme to type is better suited for this. I am sure same with GCS Wdyt?

---

this property was targeting exactly that, the scheme. The goal is to prefix tablelocation with this scheme instead of using `fs.makeQualified()`, if s3 allows multiple schemes, the OH server will select one of them. let me know if we should adapt this. cc: @jainlavina  

Type information is already collected, in `StorageProperties.getTypes().getKeySet()`

---

That's true, we have the type information available `StorageProperties.getTypes().getKeySet()`

One more question, what is the value on scheme here. If you look at s3 itself, there are various ways in which users could author this:

- https://s3.amazonaws.com/[bucket-name]/[object-key]
- - scheme: ?
- - endpoint: https://s3.amazonaws.com/
- - rootpath: bucket-name

- https://[bucket-name].s3.amazonaws.com/[object-key]
- - scheme: ?
- - endpoint: https://[bucket-name].s3.amazonaws.com/
- - rootpath: ''

- https://[bucket-name].s3-[region].amazonaws.com/[object-key]
- - scheme: ?
- - endpoint: https://[bucket-name].s3-[region].amazonaws.com
- - rootpath: ''

- https://[access-point-name]-[account-id].s3-accesspoint.[region].amazonaws.com/[object-key]
- - scheme: ?
- - endpoint: https://[access-point-name]-[account-id].s3-accesspoint.[region].amazonaws.com/
- - rootpath: ''

- s3://[bucket-name]/[object-key]
- - scheme: ?
- - endpoint: s3://
- - rootpath: bucket-name

In Hadoop Filesystem client, used against HDFS
- hdfs:///<cluster-name>/[dir-name]/[object-key]
- - scheme: ?
- - endpoint: hdfs:///<cluster-name>
- - rootpath: bucket-name

In Hadoop Filesystem client, used against Local File System
- file:///[dir-name]/[object-key]
- - scheme: ?
- - endpoint: file:///
- - rootpath: dir-name

In Hadoop Filesystem client, used against S3
- s3a:///[bucket-name]/[object-key]
- - scheme: ?
- - endpoint: s3a:///
- - rootpath: bucket-name

;tldr it is seems like endpoint could encompass scheme in pretty much all the cases. Maybe we don't need scheme, because endpoint + rootpath is sufficient?

---

great input Sumedh! and thanks for putting together various scenarios, very helpful.

I have an example where endpoint + rootpath is not sufficient and [thats for minio](https://tabular.io/blog/minio/):
there, warehouse link is: `s3://warehouse` and endpoint is `http://minio:9000`.

how do you feel about replacing scheme with `schemePrefix`, and we'll construct tableLocation by blindly adding schemePrefix+locationPath. It'll give us flexibility. 

Few Examples:

1. https://[bucket-name].s3-[region].amazonaws.com/[object-key]
schemePrefix=https://bucket-name.s3-region.amazonaws.com/
locationPath=object-key

2. s3://[bucket-name]/[object-key]
schemePrefix=s3://bucket-name/
locationPath=object-key
or 
schemePrefix=s3://
locationPath=bucket-name/object-key

3. hdfs:///[dir-name]/[object-key]
schemePrefix=hdfs:///
locationPath=bucket-name/object-key

---

>> I have an example where endpoint + rootpath is not sufficient and [thats for minio](https://tabular.io/blog/minio/):
there, warehouse link is: s3://warehouse and endpoint is http://minio:9000.

Given we will be using S3 client with MinIO,  the scheme part seems redundant and specific to how the demo catalog decided to configures it.
```
s3_client = boto3.client('s3',
                         endpoint_url=minio_endpoint,
                         aws_access_key_id=access_key,
                         aws_secret_access_key=secret_key)
s3_client.create_bucket(Bucket='my-new-bucket')
```

In our case, we would use following for minIO
```
rootPath=warehouse
endpoint=http://minio:9000
```

>> re: schemePrefix / locationPath

Let's us the terminology "endpoint", to be consistent https://docs.aws.amazon.com/general/latest/gr/s3.html 
What you have is clear, but just because this is part of cluster spec, I was thinking of start with minimum knobs. We start with endpoint and rootpath, and add scheme when we really need it.

---

Makes sense! ty resolved